src/main/java/bio/terra/service/filedata/azure/tables/TableDirectoryDao.java

```diff
@@ -0,0 +1,241 @@
+    // existing one
+    // We will create the ROOT_DIR_NAME directory here if it does not exist.
+    String lookupDirPath = fileMetadataUtils.makeLookupPath(createEntry.getPath());
+    for (String testPath = lookupDirPath;
+        !testPath.isEmpty();
+        testPath = fileMetadataUtils.getDirectoryPath(testPath)) {
+
+      // !!! In this case we are using a lookup path
+      if (lookupByFilePath(tableServiceClient, testPath) != null) {
+        break;
+      }
+
+      FireStoreDirectoryEntry dirToCreate = fileMetadataUtils.makeDirectoryEntry(testPath);
+      String rowKey = fileMetadataUtils.encodePathAsFirestoreDocumentName(testPath);
+      TableEntity entity =
+          FireStoreDirectoryEntry.toTableEntity(PARTITION_KEY, rowKey, dirToCreate);
+      TableTransactionAction t =
+          new TableTransactionAction(TableTransactionActionType.CREATE, entity);
+      createList.add(t);
+    }
+
+    String lookupPath = fileMetadataUtils.makeLookupPath(createEntry.getPath());
+    String rowKey = fileMetadataUtils.encodePathAsFirestoreDocumentName(lookupPath);
+    TableEntity createEntryEntity =
+        FireStoreDirectoryEntry.toTableEntity(PARTITION_KEY, rowKey, createEntry);
+    TableTransactionAction createEntryTransaction =
+        new TableTransactionAction(TableTransactionActionType.CREATE, createEntryEntity);
+    createList.add(createEntryTransaction);
+
+    logger.info("Creating file directory for table {}", TABLE_NAME);
+    tableClient.submitTransaction(createList);
+  }
+
+  // true - directory entry existed and was deleted; false - directory entry did not exist
+  public boolean deleteDirectoryEntry(TableServiceClient tableServiceClient, String fileId) {
+    TableClient tableClient = tableServiceClient.getTableClient(TABLE_NAME);
+
+    // Look up the directory entry by id. If it doesn't exist, we're done
+    TableEntity leafEntity = lookupByFileId(tableServiceClient, fileId);
+    if (leafEntity == null) {
+      return false;
+    }
+
+    List<TableTransactionAction> deleteList = new ArrayList<>();
+    TableTransactionAction t =
+        new TableTransactionAction(TableTransactionActionType.DELETE, leafEntity);
+    deleteList.add(t);
+
+    FireStoreDirectoryEntry leafEntry = FireStoreDirectoryEntry.fromTableEntity(leafEntity);
+    String lookupPath = fileMetadataUtils.makeLookupPath(leafEntry.getPath());
+    while (!lookupPath.isEmpty()) {
+      // Count the number of entries with this path as their directory path
+      // A value of 1 means that the directory will be empty after its child is
+      // deleted, so we should delete it also.
+      ListEntitiesOptions options =
+          new ListEntitiesOptions()
+              .setFilter(
+                  String.format(
+                      "path eq '%s'", fileMetadataUtils.makePathFromLookupPath(lookupPath)));
+      PagedIterable<TableEntity> entities = tableClient.listEntities(options, null, null);
```

Comment:
Here it seems like we are scanning the whole table. Again, having a partition key to scope/limit the scan is preferred.

Reply:
I wonder if there are perf numbers on a scan vs. partition size?